docs/input.md

```diff
@@ -47,3 +47,131 @@ itemDetailFormula
     
   }
 ```
+
+## Equality
+Formula uses input equality to determine if it should re-evaluate. A parent can cause 
+a child formula to re-evaluate by changing the input it passes. This will also trigger
+`Formula.onInputChanged` callback on the child formula.
+
+This is a desired behavior as we do want the child to react when the data that we pass changes.
+```kotlin
+data class ItemInput(
+    val itemId: String
+)
+```
+
+Making `Input` a data class and passing data as part of its properties makes it easy to reason
+about its equality. In some cases though, we want to pass objects that don't have property based
+equality such as callback functions or observables. 
+
+### Maintaining callback equality
+In many cases we want to pass callbacks to listen to formula events. 
+
+```kotlin
+data class ItemListInput(
+    val onItemSelected: (Item) -> Unit
+)
+```
+
+Formula provides an easy way to maintain callback equality. Within your parent formula, 
+you can use `FormulaContext` to instantiate callbacks by using:
+
+- `FormulaContext.callback`
+- `FormulaContext.eventCallback` 
+
+**Don't:** Don't instantiate functions within `Formula.evaluate`.
+```kotlin
+val itemListInput = ItemListInput(
+    onItemSelected = {
+        analytics.track("item_selected")
+    }
+)
+```
+
+**Do:** Use `eventCallback`
+```kotlin
+val itemListInput = ItemListInput(
+    onItemSelected = context.eventCallback {
+        transition { analytics.track("item_selected") }
+    }
+)
+```
+
+**Do:** Use already constructed callbacks
+```kotlin
+// Router is constructed outside of the "evaluate" function block
+val router: ItemRouter = ...
+
+override fun evaluate(...): ... {
+    val itemListInput = ItemListInput(
+        onItemSelected = router::onItemSelected
+    )
+}
+```
+
+**Do:** Delegate to parent input directly
+```kotlin
+override fun evaluate(input: Input, ...): ... {
+    val itemListInput = ItemListInput(
+        onItemSelected = input.onItemSelected
+    )
+}
+```
+
+### Passing observables
+Observables have identity equality which make maintaining input equality a bit tricky.
+```kotlin
+data class MyInput(
+  val eventObservable: Observable<Event>
+)
+```
+
+**Don't:** create a new observable within `Formula.evaluate`
+```kotlin
+val input = MyInput(
+    eventObservable = relay.map { Event() }
+)
+```
```

Comment:
I don't think this is the correct example - this isn't creating a new observable within `evaluate`

Reply:
Should be:
```
override fun evaluate(...) {
    val input = MyInput(
        eventObservable = eventObservable
    )
}
```